src/lib/io/sol-iio.c

```diff
@@ -1065,68 +1067,94 @@ read_channel_scan_info(struct sol_iio_channel *channel)
     return false;
 }
 
-static void
-channel_get_scale(struct sol_iio_channel *channel)
+static int
+get_scale(struct sol_iio_device *device, const char *prefix_name, double *scale)
 {
     char path[PATH_MAX];
-    char *channel_pure_name = NULL;
-    struct sol_iio_device *device = channel->device;
-    bool r;
+    char *pure_name = NULL;
+    int r;
+
+    SOL_NULL_CHECK(device, -EINVAL);
+    SOL_NULL_CHECK(prefix_name, -EINVAL);
+    SOL_NULL_CHECK(scale, -EINVAL);
 
 #define GET_SCALE(_name) \
     do { \
-        r = craft_filename_path(path, sizeof(path), CHANNEL_SCALE_PATH, \
-            device->device_id, _name); \
-        if (r && sol_util_read_file(path, "%lf", &channel->scale) > 0) { \
+        if (craft_filename_path(path, sizeof(path), CHANNEL_SCALE_PATH, \
+            device->device_id, _name)) { \
+            r = sol_util_read_file(path, "%lf", scale); \
             goto end; \
         } \
     } while (0)
 
-    GET_SCALE(channel->name);
-    /* No scale. If channel has x, y, z or Y component, look for scale file without it */
-    channel_pure_name = channel_get_pure_name(channel);
-    GET_SCALE(channel_pure_name);
+    GET_SCALE(prefix_name);
+    pure_name = channel_get_pure_name(prefix_name);
+    GET_SCALE(pure_name);
 
     SOL_INF("Could not get scale for channel [%s] in device%d. Assuming 1.0",
-        channel->name, device->device_id);
-    channel->scale = 1.0;
+        prefix_name, device->device_id);
+    *scale = 1.0;
 
 #undef GET_SCALE
 
 end:
-    free(channel_pure_name);
+    free(pure_name);
```

Comment:
if L1090 GET_SCALE goto end which pure_name is NULL. This line will crash? Should add checking for not null?